File: lab4/python/list_map.py
from typing import List, Tuple, Optional, Iterator, Type

from simple_map import SimpleMap, Key, Value


class ListMap(SimpleMap[Key, Value]):
    """
    SimpleMap implemented as an unsorted list.
    This is a very stupid idea and should not be used in practice!
    """
    elements: List[Tuple[Key, Value]]

    def __init__(self, defaultValueSupplier: Optional[Type[Value]] = None):
        super().__init__(defaultValueSupplier)
        self.elements = []

    def size(self) -> int:
        return len(self.elements)
    
    def get(self, key: Key) -> Value:
        assert key is not None, "argument must not be None"
        for key_, value in self.elements:
            if key == key_:
                return value
        # If the key does not exist, create and return a default value:
        value = self.defaultValueSupplier()
        if value is not None:
            self.elements.append((key, value))
        return value
    
    def put(self, key: Key, value: Value):
        assert key is not None, "argument must not be None"
        for i, (key_, _) in enumerate(self.elements):
            if key == key_:
                self.elements[i] = (key, value)
                return
        self.elements.append((key, value))

    def clear(self):
        self.elements = []

    def __iter__(self) -> Iterator[Key]:
        for key, _ in self.elements:
            yield key

    def __str__(self) -> str:
        if self.isEmpty(): return "ListMap (empty)"
        return f"ListMap (size {self.size()})"

    ###########################################################################
    # Validation

    def validate(self):
        duplicateKeys = self.size() != len({key for (key, _) in self.elements})
        if duplicateKeys: 
            raise ValueError(f"List map contains duplicate keys: {self}")

File: lab4/java/ListMap.java
import java.util.ArrayList;
import java.util.Iterator;
import java.util.function.Supplier;

/**
 * SimpleMap implemented as an unsorted list.
 * This is a very stupid idea and should not be used in practice!
 */
public class ListMap<Key, Value> extends SimpleMap<Key, Value> {
    ArrayList<KVPair> elements;

    ListMap() {
        this(() -> null);
    }

    ListMap(Supplier<Value> defaultValueSupplier) {
        this.elements = new ArrayList<>();
        this.defaultValueSupplier = defaultValueSupplier;
    }

    @Override
    public int size() {
        return this.elements.size();
    }

    @Override
    public Value get(Key key) {
        if (key == null) throw new NullPointerException("argument must not be null");
        for (KVPair pair : this.elements) {
            if (pair.key.equals(key)) {
                return pair.value;
            }
        }
        // If the key does not exist, generate a default value and associate with the key:
        Value value = this.defaultValueSupplier.get();
        if (value != null)
            this.elements.add(new KVPair(key, value));
        return value;
    }

    @Override
    public void put(Key key, Value value) {
        if (key == null) throw new NullPointerException("argument must not be null");
        for (KVPair pair : this.elements) {
            if (pair.key.equals(key)) {
                pair.value = value;
                return;
            }
        }
        this.elements.add(new KVPair(key, value));
    }

    @Override
    public void clear() {
        this.elements = new ArrayList<>();
    }

    @Override
    public Iterator<Key> iterator() {
        return this.elements.stream().map((p) -> p.key).iterator();
    }

    @Override
    public String toString() {
        if (this.isEmpty()) return "ListMap (empty)";
        return String.format("ListMap (size %d)", this.size());
    }

    ///////////////////////////////////////////////////////////////////////////
    // Validation

    @Override
    public void validate() {
        boolean duplicateKeys = this.size() != this.elements.stream().distinct().count();
        if (duplicateKeys)
            throw new IllegalArgumentException(String.format("List map contains duplicate keys: %s", this));
    };

    ///////////////////////////////////////////////////////////////////////////
    // Simple key-value pairs, only for internal use

    private class KVPair {
        Key key;
        Value value;
        KVPair(Key key, Value value) {
            this.key = key;
            this.value = value;
        }
    }

}

